### PR TITLE
Maven classifier support

### PR DESCRIPTION
**Pull request by cy6erGn0m**
- MavenId classifier property
- code cleanup
- classifier-aware dependency resolution 

Fixes Issue #30

### CODE REVIEW

File: src/main/kotlin/com/beust/kobalt/Main.kt [hunk 1]
```diff
@@ -91,16 +85,18 @@ private class Main @Inject constructor(
...
-            try {
-                result = runWithArgs(jc, args, argv)
-            } catch(ex: KobaltException) {
-                error(ex.message ?: "", ex)
-                result = 1
-            } finally {
-                executors.shutdown()
-            }
-        })
+        val seconds = "build".let {
```
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/Main.kt]: I don't see `it` ever used in the `let` block.

File: src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt [hunk 2]
```diff
@@ -39,19 +36,13 @@ public class DependencyManager @Inject constructor(val executors: KobaltExecutor
...
-        dependencies.forEach {
-            map.put(it.shortId, it)
-        }
-        for (k in map.keySet()) {
-            val l = map.get(k)
-            Collections.sort(l, Collections.reverseOrder())
-            result.add(l.get(0))
-        }
-        return result
+        return dependencies.groupBy { it.shortId + ":" + it.classifierIfAvailable() }
```
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt]: It's not clear to me that this code is equivalent to the one deleted, can you add a comment? A test would be nice also, this is a very core piece of functionality.

File: src/main/kotlin/com/beust/kobalt/maven/Kurl.kt [hunk 3]
```diff
@@ -1,17 +1,15 @@
...
-import javax.inject.Inject
+import java.net.*
+import javax.inject.*
 
 /**
  * Abstracts a URL so that it works transparently on either http:// or file://
  */
 class Kurl @Inject constructor(@Assisted val url: String) {
-    val connection : URLConnection by lazy {
+    val connection : URLConnection by lazy { // TODO bug here: it is impossible to close connection
```
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/Kurl.kt]: Good point.

File: src/main/kotlin/com/beust/kobalt/maven/MavenId.kt [hunk 4]
```diff
@@ -10,43 +10,52 @@ package com.beust.kobalt.maven
...
+        get() = version != null
+
+    val id: String
+        get() = MavenId.toId(groupId, artifactId, packaging, version, classifier)
 
-    private fun isVersion(s: String) : Boolean = Character.isDigit(s[0])
+}
 
-    val hasVersion = version != null
+fun MavenId(id: String): MavenId {
```
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/MavenId.kt]: I think it's confusing to create a capitalized function that acts as a second constructor. Why not move this back into `MavenId` as the second constructor, as it was?

File: src/main/kotlin/com/beust/kobalt/maven/MavenId.kt [hunk 5]
```diff
@@ -10,43 +10,52 @@ package com.beust.kobalt.maven
...
+    val c = id.split(":").toLinkedList()
 
-    val toId = MavenId.toId(groupId, artifactId, packaging, version)
+    if (c.size !in 2 .. 5) {
```
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/MavenId.kt]: We really need tests for all the combinations that a `MavenId` can have now, this is becoming hard to follow.

File: src/main/kotlin/com/beust/kobalt/maven/PomGenerator.kt [hunk 6]
```diff
@@ -55,7 +52,7 @@ public class PomGenerator @Inject constructor(@Assisted val project: Project) {
 
         val buildDir = KFiles.makeDir(project.directory, project.buildDirectory!!)
         val outputDir = KFiles.makeDir(buildDir.path, "libs")
-        val mavenId = MavenId.create(project.group!!, project.artifactId!!, project.packaging, project.version!!)
+        val mavenId = MavenId(project.group!!, project.artifactId!!, project.version!!, project.packaging)
```
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/PomGenerator.kt]: Wouldn't it make more sense to keep the arguments of `MavenId` in the same order they appear in the version?

`com.android:x:aar:1.0:packaging` means the order should be `groupId:artifactId:packaging:version:packaging`

**Comment by cbeust** [comment]
Looks great overall! Just a few comments and I'll be happy to merge.

Are you on Slack? If not, you should join us on #kobalt and ping me so we can chat.

Thanks again for all your work!

File: src/main/kotlin/com/beust/kobalt/Main.kt (hunk 1 above)
Comment (cy6erGn0m) [inline src/main/kotlin/com/beust/kobalt/Main.kt]: Yes, you are right. It's produced by a deprecated intention, will remove

File: src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt (hunk 2 above)
Comment (cy6erGn0m) [inline src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt]: No, it is not equivalent. First of all notice that `Collections.sort(l, Collections.reverseOrder())` in fact is not correct. Perhaps it should be compilation error at this place. l is a list of `IClasspathDependency` that is not comparable and it is not clear what is going to happen if there are both maven and file dependencies. 

Also now grouping key consist of shortId and classifier (if exist).
In fact it need to be tested but both maven and file dependency classes requires too much infrastrucuture. However I will try

File: src/main/kotlin/com/beust/kobalt/maven/MavenId.kt (hunk 4 above)
Comment (cy6erGn0m) [inline src/main/kotlin/com/beust/kobalt/maven/MavenId.kt]: it couldn't be moved to a secondary constructor because there is a lot of logic that couldn't fit in one expression

File: src/main/kotlin/com/beust/kobalt/maven/MavenId.kt (hunk 5 above)
Comment (cy6erGn0m) [inline src/main/kotlin/com/beust/kobalt/maven/MavenId.kt]: Don't we have already in MaveiIdTest?

File: src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt (hunk 2 above)
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt]: Then maybe let's remove this from the PR so you can look into it separately
if you want?

Overall, I think this PR could have been broken in smaller different PR's,
which would make merging them individually much easier. Just a thought for
your future PR's :)



-- 
Cédric

File: src/main/kotlin/com/beust/kobalt/maven/MavenId.kt (hunk 5 above)
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/MavenId.kt]: Yes, and I saw that you updated it with tests for the new code so all good (I hadn't gotten that far when I wrote my comment).

File: src/main/kotlin/com/beust/kobalt/maven/PomGenerator.kt (hunk 6 above)
Comment (cy6erGn0m) [inline src/main/kotlin/com/beust/kobalt/maven/PomGenerator.kt]: This order is better for manual creation because of default arguments: it is better to have optional arguments at the end

File: src/main/kotlin/com/beust/kobalt/maven/MavenId.kt (hunk 4 above)
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/MavenId.kt]: How is that a problem? Secondary constructors can be more than one expression.

File: src/main/kotlin/com/beust/kobalt/maven/PomGenerator.kt (hunk 6 above)
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/PomGenerator.kt]: But it's not a requirement. Since all these arguments are strings, I think there is too much risk to accidentally pass them in the wrong order.

File: src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt (hunk 2 above)
Comment (cy6erGn0m) [inline src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt]: It is impossible to revert this because it is *required* for valid resolution. Will try to mock all this stuff that is required for dependency instance creation.

File: src/main/kotlin/com/beust/kobalt/maven/MavenId.kt (hunk 4 above)
Comment (cy6erGn0m) [inline src/main/kotlin/com/beust/kobalt/maven/MavenId.kt]: See 

```kotlin
constructor(id: String) : this( id.parse().groupId, id.parse().artifactId, ... ) 
```

that doesn't look efficient. I believe it's not a problem to have function that looks like constructor. It's not a bug, it's a feature :)

File: src/main/kotlin/com/beust/kobalt/maven/PomGenerator.kt (hunk 6 above)
Comment (cy6erGn0m) [inline src/main/kotlin/com/beust/kobalt/maven/PomGenerator.kt]: ok, I see your point, will change signature

File: src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt (hunk 2 above)
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt]: Ok.

File: src/main/kotlin/com/beust/kobalt/maven/MavenId.kt (hunk 4 above)
Comment (cbeust) [inline src/main/kotlin/com/beust/kobalt/maven/MavenId.kt]: That's an interesting way of looking at it, but alright :)

**Comment by cbeust** [comment]
Just checking: are you still working on that PR?
